core/utils/STree.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Optional
+
+
+@dataclass
+class Node:
+    s_link: Optional[Node] = None
+    children: dict[str, Node] = field(default_factory=dict)
+    start: int = -1
+    depth: int = -1
+    parent: Optional[Node] = None
+
+    def __str__(self) -> str:
+        return f"children={list(self.children.keys())}"
+
+
+def create_node(s: str, u: Node, depth: int) -> Node:
+    start = u.start
+    parent = u.parent
+    v = Node()
+    v.start = start
+    v.depth = depth
+
+    v.children[s[start + depth]] = u
+    u.parent = v
+
+    parent.children[s[start + parent.depth]] = v
+    v.parent = parent
+
+    return v
+
+
+def create_and_attach_leaf(s: str, i: int, u: Node, depth: int) -> None:
```

Comment:
If you are inserting the leaf for suffix `i` here, then I still have no idea what the `start` and `depth` attributes are. Suffix `i` starts at `i` and goes to the end, so `len(s)-i` is the length of suffix `i`, but we don't usually store that information on an edge. With that information, we have no knowledge of the actual edge going down to leaf `i`. If `depth` is how far down the tree you insert the node, so the path-depth of node `u`, then `s[i+depth]` is the correct letter on the out-edge, though, so that part seems okay.

Reply:
You're again absolutely correct about the `depth` being how far are we doing(sum(edge_lengths) towards the current node or leaf in this case)